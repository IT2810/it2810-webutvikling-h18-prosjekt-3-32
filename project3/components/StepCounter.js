/*
Solutions is based on Pedometer API from Expo

Apple iPhone:
Data is fetched from Apple APIs and should be available instantly

Android:
Data is fetched from Google Fit APIs and requires the Google Fit app
and a successful OAuth verification to your Google account.

Boiler code from Expo docs: https://docs.expo.io/versions/latest/sdk/pedometer
*/

import Expo from 'expo';
import React from 'react';
import { Pedometer } from 'expo';
import { Text, View } from 'react-native';
import styles from "../stylesheets/StepCounter.style.js";


export default class StepCounter extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            isPedometerAvailable: "?",
            steps_last24hours: -1,
            currentStepCount: -1
        }
    }

    componentDidMount() {
        this._subscribe();
    }

    componentWillUnmount() {
        this._unsubscribe();
    }

    _subscribe = () => {
        this._subscription = Pedometer.watchStepCount(result => {
            initialSteps = this.state.steps_last24hours;
            this.setState({
                currentStepCount: initialSteps + result.steps
            });
        });

        Pedometer.isAvailableAsync().then(
            result => {
                this.setState({
                    isPedometerAvailable: String(result)
                });
            },
            error => {
                this.setState({
                    isPedometerAvailable: "Pedometer seems unavaiable, error: " + error
                });
            }
        );

        const end = new Date();
        const start = new Date();
        start.setDate(end.getDate() - 1);

        Pedometer.getStepCountAsync(start, end).then(
            result => {
                this.setState({
                    steps_last24hours: result.steps,
                    currentStepCount: result.steps
                });
            },
            error => {
                this.setState({
                    steps_last24hours: "Could not get stepCount: " + error
                });
            }
        );
    };

    _unsubscribe = () => {
        this._subscription.remove();
        this._subscription = null;
    };


    render() {
        return (
            <View style={styles.pedometerContainer}>
                <View style={styles.yellowBubble}>
                    <Text style={styles.pedometerText}>
                        Number of steps last 24 hours:
                    </Text>

                    <Text style={styles.pedometerNumber}>
                        {this.state.currentStepCount}
                    </Text>
                </View>
            </View>
    );
  }
};
