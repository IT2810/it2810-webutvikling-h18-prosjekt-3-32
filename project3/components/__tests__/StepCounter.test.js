/*
Test that components mounts as expected
Also testing that it is possible to set a new goal 
using the _storeGoal method.
*/

// Imports necessary libraries
import React from 'react';
import StepCounter from '../StepCounter';
import renderer from 'react-test-renderer';

// Tries to render new StepCounter component
test('StepCounter.js renders correctly?', () => {
    const stepTree = renderer.create(<StepCounter/>);
    expect(stepTree).toMatchSnapshot();
});

// Tries to set a new goal and compares with expected state
test("new goal appears in app", () => {
    const testStepCounterComponent = <StepCounter/>;
    const stepCounterComponent = renderer.create(testStepCounterComponent).getInstance();
    stepCounterComponent.setState({
        // Sets random goalSteps
        goalSteps: 1234
    });

    // Sets mock state
    const testList = [{
        isPedometerAvailable: "?",
        stepsToday: 0,
        // Steps since opening the app
        currentStepCount: 0,
        // State below should be updated from AsyncStorage
        goalSteps: 15000,
        // Will be updated from API
        averageStepsLastWeek: 500,
        // Will be updated based on whether goal is reached or not
        motivationalMessage: "Just a bit of walking and you will reach your goal!"
    }];

    stepCounterComponent._storeGoal(15000);

    expect(stepCounterComponent.state.goalSteps).toEqual(testList[0].goalSteps);
});