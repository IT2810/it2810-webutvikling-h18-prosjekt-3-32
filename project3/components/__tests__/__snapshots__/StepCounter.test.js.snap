// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepCounter.js renders correctly? 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "lightgray",
      "flex": 1,
      "height": "100%",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "yellow",
        "borderRadius": 10,
        "marginBottom": 20,
        "paddingBottom": 20,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 10,
        "width": "90%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Steps today:
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "paddingTop": 15,
          "textAlign": "center",
        }
      }
    >
      0
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "yellow",
        "borderRadius": 10,
        "marginBottom": 20,
        "paddingBottom": 20,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 10,
        "width": "90%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Your daily step goal:
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "paddingTop": 15,
          "textAlign": "center",
        }
      }
    >
      0
    </Text>
    <View
      animationType="timing"
      onLayout={[Function]}
      onValueChange={[Function]}
      step={500}
      style={
        Array [
          Object {
            "height": 40,
            "justifyContent": "center",
          },
          false,
          undefined,
        ]
      }
      thumbTouchSize={
        Object {
          "height": 40,
          "width": 40,
        }
      }
      value={0}
    >
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "#b3b3b3",
            },
            Object {
              "borderRadius": 2,
              "height": 4,
            },
            undefined,
          ]
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "#3f3f3f",
            "borderRadius": 2,
            "height": 4,
            "marginTop": -0,
            "opacity": 0,
            "position": "absolute",
            "width": 0,
          }
        }
      />
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "#4fcfff",
            "borderRadius": 10,
            "height": 20,
            "opacity": 0,
            "position": "absolute",
            "top": 22,
            "transform": Array [
              Object {
                "translateX": 0,
              },
              Object {
                "translateY": -0,
              },
            ],
            "width": 20,
          }
        }
      />
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {},
          ]
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "yellow",
        "borderRadius": 10,
        "marginBottom": 20,
        "paddingBottom": 20,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 10,
        "width": "90%",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "textAlign": "center",
        }
      }
    >
      Average daily steps last week:
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "paddingTop": 15,
          "textAlign": "center",
        }
      }
    >
      500
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 18,
          "fontWeight": "bold",
          "paddingTop": 15,
          "textAlign": "center",
        }
      }
    >
      Just a little more walking and you should reach your goal!
    </Text>
  </View>
</View>
`;
